File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/datastructures/Problem17413.java
package com.keunwon.algorithm.datastructures;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Problem17413 {

    public static void main(String[] args) throws IOException {
        final BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));
        final String input = br.readLine() + "\n";

        boolean isCheck = false;
        StringBuilder sb = new StringBuilder();
        Stack<Character> stack = new Stack<>();
        for (char c : input.toCharArray()) {

            if (c == ' ' || c == '\n') {
                while (!stack.isEmpty()) {
                    sb.append(stack.pop());
                }
                sb.append(c);
            } else if (c == '<') {
                while (!stack.isEmpty()) {
                    sb.append(stack.pop());
                }

                isCheck = true;
                sb.append('<');
            } else if (c == '>') {
                isCheck = false;
                sb.append('>');
            } else {
                if (isCheck) {
                    sb.append(c);
                } else {
                    stack.push(c);
                }
            }
        }

        System.out.println(sb.toString());
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem17413.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 단어 뒤집기 2
 * Level: 실버-3
 **/
class Problem17413 {
    fun solution(text: String): String {
        val stack = Stack<Char>()
        val answer = StringBuilder()
        var flag = false

        for (c in text) {
            if (c == ' ' || c == '\n') {
                while (stack.isNotEmpty()) answer.append(stack.pop())
                answer.append(' ')
            } else if (c == '<') {
                while (stack.isNotEmpty()) answer.append(stack.pop())
                answer.append('<')
                flag = true
            } else if (c == '>') {
                answer.append('>')
                flag = false
            } else {
                if (flag) answer.append(c)
                else stack.push(c)
            }
        }

        while (stack.isNotEmpty()) answer.append(stack.pop())
        return answer.toString()
    }
}

fun main() {
    val text = readLine()!!
    Problem17413().solution(text).also(::println)
}
